src/main/scala/org/constellation/domain/healthcheck/HealthCheckConsensus.scala

```diff
@@ -0,0 +1,569 @@
+package org.constellation.domain.healthcheck
+
+import cats.Order
+import cats.data.NonEmptySet
+import cats.effect.{Blocker, Clock, Concurrent, ContextShift, Fiber}
+import cats.effect.concurrent.Ref
+import cats.syntax.all._
+import io.chrisdavenport.log4cats.SelfAwareStructuredLogger
+import io.chrisdavenport.log4cats.slf4j.Slf4jLogger
+import io.circe.{Codec, Decoder, Encoder}
+import io.circe.generic.semiauto.deriveCodec
+import io.circe.syntax.EncoderOps
+import org.constellation.ConstellationExecutionContext.createSemaphore
+import org.constellation.domain.healthcheck.HealthCheckConsensusManager.{
+  FetchProposalError,
+  FetchingFailed,
+  HistoricalRound,
+  SendProposalError,
+  SendingFailed
+}
+import org.constellation.domain.p2p.PeerHealthCheck.{
+  PeerAvailable,
+  PeerHealthCheckStatus,
+  PeerUnresponsive,
+  UnknownPeer
+}
+import org.constellation.infrastructure.p2p.{ClientInterpreter, PeerResponse}
+import org.constellation.p2p.PeerData
+import org.constellation.schema.{Id, NodeState}
+import org.constellation.schema.consensus.RoundId
+
+import scala.concurrent.duration.SECONDS
+import scala.util.{Random, Try}
+
+class HealthCheckConsensus[F[_]](
+  checkedId: Id,
+  ownId: Id,
+  roundId: RoundId,
+  initialRoundPeers: Map[Id, PeerData],
+  healthCheckStatus: PeerHealthCheckStatus,
+  startedAtSecond: Long,
+  val isKeepUpRound: Boolean
+)(
+  apiClient: ClientInterpreter[F],
+  unboundedBlocker: Blocker
+)(implicit F: Concurrent[F], CS: ContextShift[F], C: Clock[F]) {
+  import HealthCheckConsensus._
+
+  implicit val roundIdOrdering: Ordering[RoundId] = Ordering.by[RoundId, String](_.id)
+  implicit val roundIdOrder: Order[RoundId] = Order.fromOrdering
+  implicit val peerDataOrdering: Ordering[PeerData] = Ordering.by[PeerData, String](_.peerMetadata.id.hex)
+  implicit val peerDataOrder: Order[PeerData] = Order.fromOrdering
+
+  val logger: SelfAwareStructuredLogger[F] = Slf4jLogger.getLogger[F]
+  val fetchProposalAfter = 120L
+  val removeUnresponsivePeersWithParallelRoundAfter = 60L
+
+  private[healthcheck] val roundIds: Ref[F, NonEmptySet[RoundId]] = Ref.unsafe(NonEmptySet.one(roundId))
+  private[healthcheck] val roundPeers: Ref[F, Map[Id, RoundData]] = {
+    Ref.unsafe(
+      initialRoundPeers
+        .mapValues(pd => RoundData(pd.asRight[ProxyPeers]))
+    )
+  }
+  private[healthcheck] val leftDuringConsensus: Ref[F, Set[Id]] = Ref.unsafe(Set.empty)
+  private[healthcheck] val removedUnresponsivePeersWithParallelRound: Ref[F, Set[Id]] = Ref.unsafe(Set.empty)
+  private[healthcheck] val ownConsensusHealthStatus: Ref[F, ConsensusHealthStatus] = Ref.unsafe(
+    ConsensusHealthStatus(
+      checkedId,
+      ownId,
+      roundId,
+      healthCheckStatus,
+      initialRoundPeers.mapValues(_.peerMetadata.nodeState)
+    )
+  )
+  private[healthcheck] val parallelRounds: Ref[F, Map[Id, Set[RoundId]]] = Ref.unsafe(Map.empty)
```

Comment:
if it's a single round `HealthCheckConsensus`, not a manager of consensus rounds, why it keeps info about other rounds in progress? shouldn't it be responsibility of a manager/supervisor of this round?

Reply:
Currently consensus is running the managements tasks related to that consensus. For parallel rounds that means that after some timeout (it's currently 60s I think) if the peer that has the parallel round is still not responding it is removed from consensus peers. This could be possibly done outside of the consensus round and stored in consensusManager but then handling that state would seem harder. Consensus is holding all the state and data upon which it is making decisions, after the consensus finishes this data is seamlessly passed on into the HistoricalRoundData class, so we should have all the info in one place.